storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provider
+
+import (
+	"math"
+	"net/url"
+
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/storage"
+
+	"github.com/juju/errors"
+	"github.com/juju/names"
+
+	"gopkg.in/goose.v1/cinder"
+	"gopkg.in/goose.v1/client"
+	"gopkg.in/goose.v1/identity"
+	"gopkg.in/goose.v1/nova"
+)
+
+const (
+	OpenstackProviderType = storage.ProviderType("openstack")
+)
+
+type openstackProvider struct {
+	clientFactory func(*storage.Config) (OpenstackAdapter, error)
+}
+
+var _ storage.Provider = (*openstackProvider)(nil)
+
+// VolumeSource implements storage.Provider.
+func (p *openstackProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
+
+	if err := p.ValidateConfig(providerConfig); err != nil {
+		return nil, err
+	}
+
+	openstackClient, err := p.clientFactory(providerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openstackVolumeSource{openstackClient}, nil
+}
+
+// FilesystemSource implements storage.Provider.
+func (p *openstackProvider) FilesystemSource(environConfig *config.Config, providerConfig *storage.Config) (storage.FilesystemSource, error) {
+	return nil, errors.NotSupportedf("filesystems")
+}
+
+// Supports implements storage.Provider.
+func (p *openstackProvider) Supports(kind storage.StorageKind) bool {
+	switch kind {
+	case storage.StorageKindBlock:
+		return true
+	}
+	return false
+}
+
+// Scope implements storage.Provider.
+func (s *openstackProvider) Scope() storage.Scope {
+	panic("not implemented")
+}
+
+// ValidateConfig implements storage.Provider.
+func (p *openstackProvider) ValidateConfig(cfg *storage.Config) error {
+
+	for _, cfgKey := range []string{
+		openstackCfgKeyUrl,
+		openstackCfgKeyUser,
+		openstackCfgKeySecrets,
+		openstackCfgKeyRegion,
+		openstackCfgKeyTenantName} {
+		if _, ok := cfg.Attrs()[cfgKey]; !ok {
+			return errors.NotAssignedf("requisite configuration was not set: %v", cfgKey)
+		}
+	}
+
+	return nil
+}
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) ([]storage.Volume, []storage.VolumeAttachment, error) {
+
+	vols := make([]storage.Volume, 0, len(args))
+	volAtt := make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+		if a.Provider != OpenstackProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				a.Provider,
+			)
+			continue
+		}
+
+		if a.Attachment != nil {
```

Comment:
Don't do the attachment here. Restrict CreateVolumes to just creating. The storage provisioner will call Attach(). We were going to remove the Attachment attribute. It may still need to be there though because with ebs for example, we need to know the AZ at creation time. Need to check the latest on this though.